### ISSUE TITLE
RFC: make JSON a builtin type

### ISSUE THREAD

**Issue by Nadrieril** [dhall-lang/dhall-lang#1027]
`Prelude.JSON.Type` has become a very important type in the Dhall ecosystem. Tools that consume Dhall are often expected to be able to handle it. However it's currently a Church-encoded user-defined type, which makes it painful for implementations to recognize and to manipulate values of that type.
Since the current focus is "displacing YAML", integrating JSON natively in Dhall makes sense to me. It would make it noticeably easier for implementations to use it, and pave the way for using it more as discussed in e.g. https://github.com/dhall-lang/dhall-lang/issues/336 and https://github.com/dhall-lang/dhall-lang/issues/121. The only required primitives would be `JSON/build` and `JSON/fold`, that would convert to and from the currently-defined Church encoding, like what `List/build` and `List/fold` do.
This would also enable a largely more debatable proposal of mine of [not normalizing under lambdas](https://discourse.dhall-lang.org/t/figuring-out-performance-bottlenecks/251/33?u=nadrieril) before caching, to prevent normal form blowup.

Transitioning to this new builtin would not be easy, since its name would conflict with the commonly used `Prelude.JSON.Type`. A partial solution could be enabling shadowing of builtin names, but that's contentious. Another option would be finding a different name for it.

**Comment by Gabriel439** [comment]
@Nadrieril: I'm in favor of this!

One way we can deal with the naming conflict is to add a `dhall lint` migration step to rename all identifiers named `JSON` to `Prelude/JSON`

**Comment by Nadrieril** [comment]
For similar reasons we could also make `Prelude.Map.Type` builtin, though that's an easier one for implementations to deal with.

**Comment by Gabriel439** [comment]
This will also make it easier to standardize a `toJSON` keyword: https://github.com/dhall-lang/dhall-lang/issues/336

**Comment by Nadrieril** [comment]
Actually I was mistaken: we would need more than just `fold` and `build` to be usable: we also need `JSON.bool`, `JSON.object`, etc. to construct values. That's annoying because then `JSON` can't directly be a builtin type, unless we add a weird special case for it. The actual type would have to be `JSON.Type`.
That reminds me of OCaml modules and the convention that if they define a type it should be `module.t`. We already assume that with the record completion operator, so maybe it makes sense for JSON too.

**Comment by philandstuff** [comment]
We already special case union types to allow selecting constructors from
them; it feels natural to do the same for JSON.

```
let Color = < Red | Green | Blue >
in Color.Red : Color
```

**Comment by Nadrieril** [comment]
@philandstuff Oh, good point!

**Comment by philandstuff** [comment]
Apologies for double-post, but a further thought occurred to me:

I wonder if we could mildly simplify the language if we were able to define JSON (and List) as special-cased recursive union types:

```
-- this Dhall is invalid because they are recursive types, but it gets the idea across
JSON = < object : Map Text JSON | array: List JSON | number : Double | text : Text | bool : Bool | null >
List = ∀ (a : Type) → < cons : { head : a, tail : List a } | nil >
```

.. so that rather than needing `List/build` and `List/fold` (and JSON equivalents), everything can be defined in terms of union constructors and `merge`.

We already do enough of this with `Optional` that we have dispensed with `Optional/build` and `Optional/fold` - we use `merge` for destruction and special-case constructors `Some` and `None` for construction.  My proposal is to do so more consistently for more builtin types.

**Comment by Nadrieril** [comment]
The reason we don't currently do this is that this would kill the totality of Dhall. The reason `List/fold` ensures totality is because it forces the function that "recursively" consumes the list to have the form of a fold. If we wanted to do the same with just `merge`, we would need to allow functions to be recursive, and we can't enforce that they terminate anymore. Now that I've said that though, there might be a way to change the behavior of merge in the presence of a recursive type to only allow folds?

**Comment by Nadrieril** [comment]
Huh, that actually seems doable. The bulk of the difficulty would be in restricting the kind of self-reference allowed when defining a type, to prevent e.g. `Foo = { x: Bool, y: Foo }` or polymorphic recursion. Once we have that though, it looks like a sufficiently clever `merge` could preserve totality: for example, `merge` on `List a` would take
```dhall
{
	nil: r,
	cons: { head : a, tail : r } -> r
}
```
i.e. like normal `merge` on a union type, but also replacing any instance of `List a` by the return type `r`. Since we disallow polymorphic recursion, we're sure that there is no instance of `List b` with `b != a`.

The idea is pretty, it would make Dhall way more uniform. At the cost of making the spec of `merge` even more hellish x). And I'm not confident that allowing self-reference when defining types fits well with Dhall, since everything is a term.
Another drawback is that implementaitons would still be expected to handle those types specifically when e.g. marshalling to native types. Having only user-defined types would make this very painful (and probably slow).

**Comment by Gabriel439** [comment]
@Nadrieril: I do think there is value in adding language support for a type-level fixed point, which would essentially be equivalent to this definition:

```dhall
let Recursive : (Type → Type) → Type =
      \(f : Type → Type) → forall (r : Type) → (f r → r) → r

in  Recursive
```

... except opaque (so that normal forms for such recursive types would still look nice), and then provide a general `fold` keyword that would essentially "unwrap" the opaque type to produce the equivalent Church encoding implementation:

```dhall
-- We could require that `fold` is fully saturated so that it could infer `f` and `r`
fold : Recursive f → (f r → r) → r
```

... and then perhaps another keyword for the special case where `f` is a sum type like you suggest.  However, I might suggest avoiding reusing the `merge` keyword because then we'd have to perform type inference during normalization to detect whether or not to trigger  this special behavior.  However, we could add another keyword for this purpose so that no guesswork is necessary.

**Comment by Nadrieril** [comment]
It depends how we handle value of type `fix(..) -> ...`. I would assume they are built with a special `Fix` constructor, that we would make invisible in the syntax. This would be like Haskell's `data Fix f = f (Fix f)` except that the constructor is hidden. So `merge` should only have to unpack that constructor automatically.
The variants of the recursive types would also invisibly wrap with this constructor: given
```
Nat = fix(nat : Type) → < zero | succ: nat >
```
then `Nat.succ x` normalizes to `fix(< zero | succ: Nat >.succ x)`

Roughly, in sequents:
```
u ⇥ fix(e)   merge t e ⇥ b
──────────────────────────
merge t u ⇥ b


e ⇥ fix(rec : Type) → u₀   u₀ ⇥ < xs… >  u₀[rec ≔ e] = u₁  fix(u₁.x v) ⇥ b
──────────────────────────────────────────────────────────────────────────
e.x v ⇥ b
```

I agree that this would be a neat addition to the language. I would still prefer List and JSON to be builtins.

**Comment by Nadrieril** [comment]
Oh wait no you're right, this merge only removes one layer but we want it to be recursive. I need to think about this more carefully

**Comment by Gabriel439** [comment]
@Nadrieril: Yeah, I agree that even if there were language support for recursion, we would still also want language support for an opaque `JSON` type

**Comment by k** [comment]
Any update on this? I like the idea of dhall but migrating over from existing helm charts currently seems impossible with no interoperability. In theory one could make a tool to merge dhall with other configuration renderers and gradually migrate that way.

Is there currently any way to import arbitrary JSON and have it typed as `any`?

**Comment by Gabriel439** [comment]
@k: There is not yet language support, but there are command-line tools for converting JSON and YAML to Dhall.  See:

* [Dhall Manual - How to convert an existing YAML configuration to Dhall](https://github.com/Gabriel439/dhall-manual/blob/e1404fffc18ee127908fc9ece664473da5d31b86/manuscript/01-YamlToDhall.md)
